Support resilient properties of move-only type

kavon opened the pull request:
resolves rdar://104504239

kavon committed:
treat move-only storage decls like `@_borrowed`

This change makes sure a _read accessor is always available for
properties of move-only type. If programmers define a `get`,
a `_read` will be synthesized to call it and then return a borrow
of the owned value.

kavon committed:
re-allow non-final classes to have move-only members

kavon committed:
avoid copying from `begin_apply` when `_read`-ing a move-only type

When an expression accessing a move-only type through a `_read`
accessor, it can only be borrowed. While a simple access like
`someClass.moveOnlyField` would get recognized in SILGen and properly
borrowed, if you did any further access off of that then you'd get an
illegal sequence involving a copy that is hard to eliminate. In
particular, if you wrote `someClass.moveOnlyField.method()` then we
would emit something like:

```
(%borrowedMO, %coro) = begin_apply #SomeClass.moveOnlyField!.read(...)
%copiedMO = copy_value %borrowedMO
end_apply %coro
= apply #MoveOnlyType.method(%copiedMO) // param is just @guaranteed
```

That's wrong, since we need to use the borrow to call the method, but
that borrow's lifetime ends at the `end_apply`.

Turns out the fix is rather subtle. The reason the access above wouldn't
work is that it did _not_ have a `LoadExpr` around it, which is what the
code in the `ArgEmitter` was expecting when doing an ad-hoc check to see
if it needs to emit as a borrow.

kavon commented:
@swift-ci please smoke test

gottesmm commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -2933,7 +2933,39 @@ static void emitDelayedArguments(SILGenFunction &SGF,
+  // The direct storage reference.
+  Expr *storageRef;
+  // The root of the expression that accesses the storage in \c storageRef.
+  Expr *transitiveRoot;
```
Can you add an example here of an AST and what would be the storage ref and what would be the transitiveRoot? Maybe add it to the top level comment on \p StorageRefResult?

gottesmm commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -2933,7 +2933,39 @@ static void emitDelayedArguments(SILGenFunction &SGF,
+};
+} // namespace
+
+static StorageRefResult findStorageReferenceExprForBorrow(Expr *e) {
```
Why not make this a static method on StorageRefResult? It feels like a static constructor.

gottesmm commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -2933,7 +2933,39 @@ static void emitDelayedArguments(SILGenFunction &SGF,
+  }
+
+  // By default, this result is the transitive root.
+  operator Expr *() const { return transitiveRoot; }
```
Can you turn this into a method with a name to make it more explicit? Maybe getTransitiveRoot()?

gottesmm commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -2956,71 +2988,102 @@ static Expr *findStorageReferenceExprForBorrow(Expr *e) {
+    sawLoad = true;
+  }
+
+  // If we're consuming instead, then the load _must_ have been there.
```
Can you add more to the comment. Why must have the load been there and what is the significance of that? Your comment here is not telling me why you are doing this. Just describing the situation.

kavon replied:
I don't have a concrete reason. There was no comment there before explaining why a load must be there.

gottesmm commented on lib/SILGen/ArgumentSource.h:
```diff
@@ -222,8 +222,15 @@ class ArgumentSource {
   }
 
+  /// The kind of operation under which we are querying a storage reference.
+  enum class StorageReferenceKind {
```
I find this to be pretty misleading as a name. Can you rename it to something with the name operation. Maybe: StorageReferenceOperationKind?

Otherwise, it sounds like you are talking about the actual kind of the StorageReference... which is not what on the face of it you are trying to talk about.

gottesmm commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -2956,71 +2988,102 @@ static Expr *findStorageReferenceExprForBorrow(Expr *e) {
 
-  auto *lvExpr = ::findStorageReferenceExprForBorrow(li->getSubExpr());
+  // NOTE: In this search, we do not look through load implying that if we have
+  // copyable types extracted from a move only class, we will not borrow.
```
I find this NOTE confusing since we looked through a load above... of course after reading more deeply, I know you are talking about a transitive load due to the class. Can you change this to include information stating that thus we won't look through a recursive load... something that would occur if we extract from a move only class. I think it will make it clearer to a reader who doesn't know this code.

gottesmm reviewed (changes requested):
Some initial comments.

kavon committed:
fix bug when devirtualizing a `begin_apply` during inlining

When devirtualizing a `begin_apply`, it was passing the token's
use list to the conversion function when trying to convert the
yielded result. It's suppose to be the yielded result's list.

This became apparent when it encountered an access of a
`@_borrowed` property and we hit an assertion about an empty
use-list of a guaranteed value, when it was in fact the wrong list!

kavon committed:
no_consume_or_assign ref_element_addr should not have any destroys

This fixes an issue when doing move-checking on a read accessor,
where the field is only borrowed. After the MoveOnlyAddressChecker
ran on it, it'd inject a destroy that didn't get "claimed":

```
  %2 = ref_element_addr %0 : $ListOfFiles, #ListOfFiles.file
  %3 = mark_must_check [no_consume_or_assign] %2 : $*File
  %4 = begin_access [read] [dynamic] %3 : $*File
  %5 = load_borrow %4 : $*File
  yield %5 : $File, resume bb1, unwind bb2

bb1:
  end_borrow %5 : $File
  end_access %4 : $*File
  destroy_addr %2 : $*File // BAD
  %9 = tuple ()
  return %9 : $()
```

The approach of this fix is to recognize that at the point we're
injecting destroys, we would have emitted diagnostics and stopped
already if there were any consuming uses that we need to clean-up
after, since we're in `no_consume_or_assign` checking mode here
when just reading the field.

kavon committed:
add SIL output test case for coroutine accessors

kavon committed:
update existing SILGen tests

Note that the changes from `load_borrow` to
`load [copy]` are for copyable types at the end
of an access chain. So far this patch just makes
more noncopyable things get borrowed when accessed,
like globals, which is what we want.

kavon commented:
@swift-ci please smoke test

kavon commented:
@swift-ci please test

kavon committed:
add a bunch of hopeful tests

kavon commented:
@swift-ci please smoke test and merge

kavon commented:
```
LD_LIBRARY_PATH=../lib:../stubdata:../tools/ctestfw:$LD_LIBRARY_PATH  ../bin/makeconv -s /home/build-user/icu/icu4c/source/data -d ./out/build/icudt65l -c mappings/ibm-1149_P100-1997.ucm
make[1]: *** [../data/rules.mk:750: out/build/icudt65l/curr/supplementalData.res] Segmentation fault (core dumped)
```

linux failure loooks random

kavon commented:
@swift-ci please smoke test and merge
